nuxeo-drive-jsf/src/main/java/org/nuxeo/drive/seam/NuxeoDriveActions.java

```diff
@@ -209,29 +215,50 @@ public String getDriveEditURL(DocumentModel currentDocument) {
+    }
+
+    public String getDriveEditURL(DocumentModel doc, String xPath) {
+        if (doc == null) {
+            return null;
+        }
+        Object obj = doc.getPropertyValue(xPath);
+        if (!(obj instanceof Blob)) {
+            throw new NuxeoException(String.format(
+                "Property %s of document %s (%s) is not a blob, cannot get Drive Edit URL.",
+                xPath, doc.getPathAsString(), doc.getId()));
+        }
+        Blob blob = (Blob) obj;
+
+        return getDriveEditURL(doc, blob, xPath);
+    }
+
+    public String getDriveEditURL(DocumentModel doc, Blob blob, String xPath) {
+        if (doc == null || blob == null) {
+            return null;
+        }
+
+        String editURL = "%s://%s/%suser/%s/repo/%s/nxdocid/%s/filename/%s/downloadUrl/%s";
         ServletRequest servletRequest = (ServletRequest) FacesContext.getCurrentInstance()
                                                                      .getExternalContext()
                                                                      .getRequest();
-        String baseURL = VirtualHostHelper.getBaseURL(servletRequest);
-        StringBuilder sb = new StringBuilder();
-        sb.append(NXDRIVE_PROTOCOL).append("://");
-        sb.append(PROTOCOL_COMMAND_EDIT).append("/");
-        sb.append(baseURL.replaceFirst("://", "/"));
-        sb.append("user/");
-        sb.append(documentManager.getPrincipal().getName());
-        sb.append("/");
-        sb.append("repo/");
-        sb.append(documentManager.getRepositoryName());
-        sb.append("/nxdocid/");
-        sb.append(currentDocument.getId());
-        sb.append("/filename/");
-        String escapedFilename = fileName.replaceAll("(/|\\\\|\\*|<|>|\\?|\"|:|\\|)", "-");
-        sb.append(URIUtils.quoteURIPathComponent(escapedFilename, true));
-        sb.append("/downloadUrl/");
+        String baseURL = VirtualHostHelper.getBaseURL(servletRequest).replaceFirst("://", "/");
+
+        String user = documentManager.getPrincipal().getName();
+        String repo = documentManager.getRepositoryName();
+        String docId = doc.getId();
+        String filename = blob.getFilename();
+        filename = filename.replaceAll("(/|\\\\|\\*|<|>|\\?|\"|:|\\|)", "-");
+        filename = URIUtils.quoteURIPathComponent(filename, true);
         DownloadService downloadService = Framework.getService(DownloadService.class);
-        String downloadUrl = downloadService.getDownloadUrl(currentDocument, DownloadService.BLOBHOLDER_0, "");
-        sb.append(downloadUrl);
+        if (xPath == null) {
+            xPath = DownloadService.BLOBHOLDER_0;
+        }
+        String downloadUrl = downloadService.getDownloadUrl(doc, xPath, filename);
+
+        StringBuilder sb = new StringBuilder();
+        Formatter fmt = new Formatter(sb);
+        fmt.format(editURL, NXDRIVE_PROTOCOL, PROTOCOL_COMMAND_EDIT, baseURL, user, repo, docId, filename, downloadUrl);
```

Comment:
Standard Java idiom is not to use a `Formatter` object, but `String.format(format, params...)`